{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "elife_test = pd.read_csv(\"eLife_test.csv\")\n",
    "plos_test = pd.read_csv(\"PLOS_test.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>article</th>\n",
       "      <th>headings</th>\n",
       "      <th>keywords</th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acylation of diverse carbohydrates occurs acro...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Results and disc...</td>\n",
       "      <td>['biochemistry and chemical biology', 'computa...</td>\n",
       "      <td>elife-81547-v1</td>\n",
       "      <td>A novel fold for acyltransferase-3 (AT3) prote...</td>\n",
       "      <td>Acylation of diverse carbohydrates occurs acro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Honey bee ecology demands they make both rapid...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Results', 'Discu...</td>\n",
       "      <td>['computational and systems biology']</td>\n",
       "      <td>elife-86176-v2</td>\n",
       "      <td>How honey bees make fast and accurate decisions</td>\n",
       "      <td>Honey bee ecology demands they make both rapid...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Biguanides , including the world’s most prescr...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Results', 'Discu...</td>\n",
       "      <td>['genetics and genomics']</td>\n",
       "      <td>elife-82210-v1</td>\n",
       "      <td>Ether lipid biosynthesis promotes lifespan ext...</td>\n",
       "      <td>Biguanides , including the world’s most prescr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ecological relationships between bacteria medi...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Results', 'Discu...</td>\n",
       "      <td>['microbiology and infectious disease', 'ecolo...</td>\n",
       "      <td>elife-83152-v2</td>\n",
       "      <td>Universal gut microbial relationships in the g...</td>\n",
       "      <td>Ecological relationships between bacteria medi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gamma oscillations are believed to underlie co...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Results', 'Discu...</td>\n",
       "      <td>['neuroscience']</td>\n",
       "      <td>elife-83044-v2</td>\n",
       "      <td>Mechanisms and functions of respiration-driven...</td>\n",
       "      <td>Gamma oscillations are believed to underlie co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>High-throughput transgenesis using synthetic D...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Results', 'Discu...</td>\n",
       "      <td>['tools and resources', 'genetics and genomics']</td>\n",
       "      <td>elife-84831-v1</td>\n",
       "      <td>High-throughput library transgenesis in Caenor...</td>\n",
       "      <td>High-throughput transgenesis using synthetic D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>To reach their final destinations , outer memb...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Results', 'Discu...</td>\n",
       "      <td>['structural biology and molecular biophysics'...</td>\n",
       "      <td>elife-82885-v1</td>\n",
       "      <td>Deep learning-driven insights into super prote...</td>\n",
       "      <td>To reach their final destinations , outer memb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>Midbrain dopamine ( DA ) neurons are key regul...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Results', 'Discu...</td>\n",
       "      <td>['neuroscience']</td>\n",
       "      <td>elife-87902-v2</td>\n",
       "      <td>Conditional deletion of neurexins dysregulates...</td>\n",
       "      <td>Midbrain dopamine ( DA ) neurons are key regul...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>Mobile health ( mHealth ) interventions , whic...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Methods', 'Resul...</td>\n",
       "      <td>['medicine', 'epidemiology and global health']</td>\n",
       "      <td>elife-79615-v2</td>\n",
       "      <td>Lack of ownership of mobile phones could hinde...</td>\n",
       "      <td>Mobile health ( mHealth ) interventions , whic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>Severe acute respiratory syndrome coronavirus ...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Methods', 'Resul...</td>\n",
       "      <td>['medicine', 'epidemiology and global health']</td>\n",
       "      <td>elife-80428-v2</td>\n",
       "      <td>Antibody levels following vaccination against ...</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>142 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               article  \\\n",
       "0    Acylation of diverse carbohydrates occurs acro...   \n",
       "1    Honey bee ecology demands they make both rapid...   \n",
       "2    Biguanides , including the world’s most prescr...   \n",
       "3    Ecological relationships between bacteria medi...   \n",
       "4    Gamma oscillations are believed to underlie co...   \n",
       "..                                                 ...   \n",
       "137  High-throughput transgenesis using synthetic D...   \n",
       "138  To reach their final destinations , outer memb...   \n",
       "139  Midbrain dopamine ( DA ) neurons are key regul...   \n",
       "140  Mobile health ( mHealth ) interventions , whic...   \n",
       "141  Severe acute respiratory syndrome coronavirus ...   \n",
       "\n",
       "                                              headings  \\\n",
       "0    ['Abstract', 'Introduction', 'Results and disc...   \n",
       "1    ['Abstract', 'Introduction', 'Results', 'Discu...   \n",
       "2    ['Abstract', 'Introduction', 'Results', 'Discu...   \n",
       "3    ['Abstract', 'Introduction', 'Results', 'Discu...   \n",
       "4    ['Abstract', 'Introduction', 'Results', 'Discu...   \n",
       "..                                                 ...   \n",
       "137  ['Abstract', 'Introduction', 'Results', 'Discu...   \n",
       "138  ['Abstract', 'Introduction', 'Results', 'Discu...   \n",
       "139  ['Abstract', 'Introduction', 'Results', 'Discu...   \n",
       "140  ['Abstract', 'Introduction', 'Methods', 'Resul...   \n",
       "141  ['Abstract', 'Introduction', 'Methods', 'Resul...   \n",
       "\n",
       "                                              keywords              id  \\\n",
       "0    ['biochemistry and chemical biology', 'computa...  elife-81547-v1   \n",
       "1                ['computational and systems biology']  elife-86176-v2   \n",
       "2                            ['genetics and genomics']  elife-82210-v1   \n",
       "3    ['microbiology and infectious disease', 'ecolo...  elife-83152-v2   \n",
       "4                                     ['neuroscience']  elife-83044-v2   \n",
       "..                                                 ...             ...   \n",
       "137   ['tools and resources', 'genetics and genomics']  elife-84831-v1   \n",
       "138  ['structural biology and molecular biophysics'...  elife-82885-v1   \n",
       "139                                   ['neuroscience']  elife-87902-v2   \n",
       "140     ['medicine', 'epidemiology and global health']  elife-79615-v2   \n",
       "141     ['medicine', 'epidemiology and global health']  elife-80428-v2   \n",
       "\n",
       "                                                 title  \\\n",
       "0    A novel fold for acyltransferase-3 (AT3) prote...   \n",
       "1      How honey bees make fast and accurate decisions   \n",
       "2    Ether lipid biosynthesis promotes lifespan ext...   \n",
       "3    Universal gut microbial relationships in the g...   \n",
       "4    Mechanisms and functions of respiration-driven...   \n",
       "..                                                 ...   \n",
       "137  High-throughput library transgenesis in Caenor...   \n",
       "138  Deep learning-driven insights into super prote...   \n",
       "139  Conditional deletion of neurexins dysregulates...   \n",
       "140  Lack of ownership of mobile phones could hinde...   \n",
       "141  Antibody levels following vaccination against ...   \n",
       "\n",
       "                                              abstract  \n",
       "0    Acylation of diverse carbohydrates occurs acro...  \n",
       "1    Honey bee ecology demands they make both rapid...  \n",
       "2    Biguanides , including the world’s most prescr...  \n",
       "3    Ecological relationships between bacteria medi...  \n",
       "4    Gamma oscillations are believed to underlie co...  \n",
       "..                                                 ...  \n",
       "137  High-throughput transgenesis using synthetic D...  \n",
       "138  To reach their final destinations , outer memb...  \n",
       "139  Midbrain dopamine ( DA ) neurons are key regul...  \n",
       "140  Mobile health ( mHealth ) interventions , whic...  \n",
       "141  Severe acute respiratory syndrome coronavirus ...  \n",
       "\n",
       "[142 rows x 6 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "elife_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_prompt(df, i):\n",
    "    # keywords = df[\"keywords\"][i]\n",
    "    title = df[\"title\"][i]\n",
    "    abstract = df[\"abstract\"][i]\n",
    "    return f\"Title: {title}\\n\\nAbstract: {abstract}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Title: A novel fold for acyltransferase-3 (AT3) proteins provides a framework for transmembrane acyl-group transfer\n",
      "\n",
      "Abstract: Acylation of diverse carbohydrates occurs across all domains of life and can be catalysed by proteins with a membrane bound acyltransferase-3 ( AT3 ) domain ( PF01757 ) . In bacteria , these proteins are essential in processes including symbiosis , resistance to viruses and antimicrobials , and biosynthesis of antibiotics , yet their structure and mechanism are largely unknown . In this study , evolutionary co-variance analysis was used to build a computational model of the structure of a bacterial O-antigen modifying acetyltransferase , OafB . The resulting structure exhibited a novel fold for the AT3 domain , which molecular dynamics simulations demonstrated is stable in the membrane . The AT3 domain contains 10 transmembrane helices arranged to form a large cytoplasmic cavity lined by residues known to be essential for function . Further molecular dynamics simulations support a model where the acyl-coA donor spans the membrane through accessing a pore created by movement of an important loop capping the inner cavity , enabling OafB to present the acetyl group close to the likely catalytic resides on the extracytoplasmic surface . Limited but important interactions with the fused SGNH domain in OafB are identified , and modelling suggests this domain is mobile and can both accept acyl-groups from the AT3 and then reach beyond the membrane to reach acceptor substrates . Together this new general model of AT3 function provides a framework for the development of inhibitors that could abrogate critical functions of bacterial pathogens . \n"
     ]
    }
   ],
   "source": [
    "print(make_prompt(elife_test,0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dachengma/miniconda3/envs/llm/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "Loading checkpoint shards: 100%|██████████| 9/9 [00:02<00:00,  3.93it/s]\n",
      "Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.\n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "Expected one of cpu, cuda, ipu, xpu, mkldnn, opengl, opencl, ideep, hip, ve, fpga, ort, xla, lazy, vulkan, mps, meta, hpu, mtia, privateuseone device type at start of device string: auto",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 9\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# model_id = \"aaditya/OpenBioLLM-Llama3-8B\"\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m#model_path = \"/home/dachengma/Llama3-OpenBioLLM-8B/\"\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;66;03m# model_path = \"/home/dachengma/models/LLM-Research/Meta-Llama-3-8B-Instruct\"\u001b[39;00m\n\u001b[1;32m      7\u001b[0m model_path\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/home/dachengma/biolaysumm2024_data/LLaMA-Factory/llama3_lora_merged_eLife\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m----> 9\u001b[0m pipeline \u001b[38;5;241m=\u001b[39m \u001b[43mtransformers\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpipeline\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     10\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtext-generation\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     11\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmodel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmodel_path\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     12\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmodel_kwargs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m{\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtorch_dtype\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mtorch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbfloat16\u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     13\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdevice\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mauto\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     14\u001b[0m \u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/transformers/pipelines/__init__.py:1108\u001b[0m, in \u001b[0;36mpipeline\u001b[0;34m(task, model, config, tokenizer, feature_extractor, image_processor, framework, revision, use_fast, token, device, device_map, torch_dtype, trust_remote_code, model_kwargs, pipeline_class, **kwargs)\u001b[0m\n\u001b[1;32m   1105\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m device \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1106\u001b[0m     kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdevice\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m device\n\u001b[0;32m-> 1108\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mpipeline_class\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mframework\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mframework\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/transformers/pipelines/text_generation.py:84\u001b[0m, in \u001b[0;36mTextGenerationPipeline.__init__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m     83\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__init__\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m---> 84\u001b[0m     \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__init__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     85\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcheck_model_type(\n\u001b[1;32m     86\u001b[0m         TF_MODEL_FOR_CAUSAL_LM_MAPPING_NAMES \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mframework \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtf\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m MODEL_FOR_CAUSAL_LM_MAPPING_NAMES\n\u001b[1;32m     87\u001b[0m     )\n\u001b[1;32m     88\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mprefix\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_preprocess_params:\n\u001b[1;32m     89\u001b[0m         \u001b[38;5;66;03m# This is very specific. The logic is quite complex and needs to be done\u001b[39;00m\n\u001b[1;32m     90\u001b[0m         \u001b[38;5;66;03m# as a \"default\".\u001b[39;00m\n\u001b[1;32m     91\u001b[0m         \u001b[38;5;66;03m# It also defines both some preprocess_kwargs and generate_kwargs\u001b[39;00m\n\u001b[1;32m     92\u001b[0m         \u001b[38;5;66;03m# which is why we cannot put them in their respective methods.\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/transformers/pipelines/base.py:855\u001b[0m, in \u001b[0;36mPipeline.__init__\u001b[0;34m(self, model, tokenizer, feature_extractor, image_processor, modelcard, framework, task, args_parser, device, torch_dtype, binary_output, **kwargs)\u001b[0m\n\u001b[1;32m    853\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mxpu\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m device \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_torch_xpu_available(check_device\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m):\n\u001b[1;32m    854\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mdevice\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m is not available, you should use device=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcpu\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m instead\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m--> 855\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdevice \u001b[38;5;241m=\u001b[39m \u001b[43mtorch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdevice\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdevice\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    856\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m device \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    857\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdevice \u001b[38;5;241m=\u001b[39m torch\u001b[38;5;241m.\u001b[39mdevice(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcpu\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Expected one of cpu, cuda, ipu, xpu, mkldnn, opengl, opencl, ideep, hip, ve, fpga, ort, xla, lazy, vulkan, mps, meta, hpu, mtia, privateuseone device type at start of device string: auto"
     ]
    }
   ],
   "source": [
    "import transformers\n",
    "import torch\n",
    "\n",
    "# model_id = \"aaditya/OpenBioLLM-Llama3-8B\"\n",
    "#model_path = \"/home/dachengma/Llama3-OpenBioLLM-8B/\"\n",
    "# model_path = \"/home/dachengma/models/LLM-Research/Meta-Llama-3-8B-Instruct\"\n",
    "model_path=\"/home/dachengma/biolaysumm2024_data/LLaMA-Factory/llama3_lora_merged_eLife\"\n",
    "\n",
    "pipeline = transformers.pipeline(\n",
    "    \"text-generation\",\n",
    "    model=model_path,\n",
    "    model_kwargs={\"torch_dtype\": torch.bfloat16},\n",
    "    device=\"cuda\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'make_top_shot_prompt' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 7\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# messages = [\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m#     {\"role\": \"system\", \"content\": \"You are an expert  from the healthcare and biomedical domain with extensive medical knowledge and practical experience. Your name is OpenBioLLM, and you were developed by Saama AI Labs. who's willing to help answer the user's query with explanation. In your explanation, leverage your deep medical expertise such as relevant anatomical structures, physiological processes, diagnostic criteria, treatment guidelines, or other pertinent medical concepts. Use precise medical terminology while still aiming to make the explanation clear and accessible to a general audience.\"},\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m#     {\"role\": \"user\", \"content\": \"How can i split a 3mg or 4mg waefin pill so i can get a 2.5mg pill?\"},\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# ]\u001b[39;00m\n\u001b[1;32m      6\u001b[0m messages \u001b[38;5;241m=\u001b[39m [\n\u001b[0;32m----> 7\u001b[0m     {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrole\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msystem\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcontent\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mYou are an expert in writing lay summaries for biomedical literature, for example, here is one example of abstract and paired lay summary \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmake_top_shot_prompt(elife_test,\u001b[38;5;250m \u001b[39m\u001b[38;5;241m0\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124myou will write lay summary in 10-20 sentences for the abstract that is given to you. write in readable way but maximize relevance and factuality. only respond the actual lay summary, here is the given title and abstract,\u001b[39m\u001b[38;5;124m\"\u001b[39m},\n\u001b[1;32m      8\u001b[0m     {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrole\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124muser\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcontent\u001b[39m\u001b[38;5;124m\"\u001b[39m: make_prompt(elife_test, \u001b[38;5;241m0\u001b[39m)}\n\u001b[1;32m      9\u001b[0m ]\n\u001b[1;32m     11\u001b[0m prompt \u001b[38;5;241m=\u001b[39m pipeline\u001b[38;5;241m.\u001b[39mtokenizer\u001b[38;5;241m.\u001b[39mapply_chat_template(\n\u001b[1;32m     12\u001b[0m         messages, \n\u001b[1;32m     13\u001b[0m         tokenize\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, \n\u001b[1;32m     14\u001b[0m         add_generation_prompt\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m     15\u001b[0m )\n\u001b[1;32m     17\u001b[0m terminators \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m     18\u001b[0m     pipeline\u001b[38;5;241m.\u001b[39mtokenizer\u001b[38;5;241m.\u001b[39meos_token_id,\n\u001b[1;32m     19\u001b[0m     pipeline\u001b[38;5;241m.\u001b[39mtokenizer\u001b[38;5;241m.\u001b[39mconvert_tokens_to_ids(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m<|eot_id|>\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     20\u001b[0m ]\n",
      "\u001b[0;31mNameError\u001b[0m: name 'make_top_shot_prompt' is not defined"
     ]
    }
   ],
   "source": [
    "\n",
    "# messages = [\n",
    "#     {\"role\": \"system\", \"content\": \"You are an expert  from the healthcare and biomedical domain with extensive medical knowledge and practical experience. Your name is OpenBioLLM, and you were developed by Saama AI Labs. who's willing to help answer the user's query with explanation. In your explanation, leverage your deep medical expertise such as relevant anatomical structures, physiological processes, diagnostic criteria, treatment guidelines, or other pertinent medical concepts. Use precise medical terminology while still aiming to make the explanation clear and accessible to a general audience.\"},\n",
    "#     {\"role\": \"user\", \"content\": \"How can i split a 3mg or 4mg waefin pill so i can get a 2.5mg pill?\"},\n",
    "# ]\n",
    "\n",
    "messages = [\n",
    "    {\"role\": \"system\", \"content\": f\"You are an expert in writing lay summaries for biomedical literature, for example, here is one example of abstract and paired lay summary {make_top_shot_prompt(elife_test, 0)}you will write lay summary in 10-20 sentences for the abstract that is given to you. write in readable way but maximize relevance and factuality. only respond the actual lay summary, here is the given title and abstract,\"},\n",
    "    {\"role\": \"user\", \"content\": make_prompt(elife_test, 0)}\n",
    "]\n",
    "\n",
    "prompt = pipeline.tokenizer.apply_chat_template(\n",
    "        messages, \n",
    "        tokenize=False, \n",
    "        add_generation_prompt=True\n",
    ")\n",
    "\n",
    "terminators = [\n",
    "    pipeline.tokenizer.eos_token_id,\n",
    "    pipeline.tokenizer.convert_tokens_to_ids(\"<|eot_id|>\")\n",
    "]\n",
    "\n",
    "outputs = pipeline(\n",
    "    prompt,\n",
    "    max_new_tokens=2048,\n",
    "    eos_token_id=terminators,\n",
    "    do_sample=True,\n",
    "    temperature=0.0001,\n",
    "    top_p=0.9,\n",
    ")\n",
    "print(outputs[0][\"generated_text\"][len(prompt):])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# KNN adaptive in-context learning \n",
    "elife_3shot = pd.read_json(\"eLife_generated_3shot_examples_bge-m3.json\", lines=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "elife_test[\"3shot\"] = elife_3shot[\"FewShotExamples\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Abstract': 'Human-associated microbial communities vary across individuals: possible contributing factors include ( genetic ) relatedness , diet , and age . However , our surroundings , including individuals with whom we interact , also likely shape our microbial communities . To quantify this microbial exchange , we surveyed fecal , oral , and skin microbiota from 60 families ( spousal units with children , dogs , both , or neither ) . Household members , particularly couples , shared more of their microbiota than individuals from different households , with stronger effects of co-habitation on skin than oral or fecal microbiota . Dog ownership significantly increased the shared skin microbiota in cohabiting adults , and dog-owning adults shared more ‘skin’ microbiota with their own dogs than with other dogs . Although the degree to which these shared microbes have a true niche on the human body , vs transient detection after direct contact , is unknown , these results suggest that direct and frequent contact with our cohabitants may significantly shape the composition of our microbial communities',\n",
       " 'Lay Summary': 'The human body is home to many different microorganisms , with a range of bacteria , fungi and archaea living on the skin , in the intestine and at various other sites in the body . While many of these microorganisms are beneficial to their human hosts , we know very little about most of them . Early research focused primarily on comparing the microorganisms found in healthy individuals with those found in individuals suffering from a particular illness . More recently researchers have become interested in more general issues , such as understanding how these collections of microorganisms , which are also known as the human microbiota or the human microbiome , become established , and exploring the causes of similarities and differences between the microbiota of individuals . We now know that the communities of microorganisms found in the intestines of genetically related people tend to be more similar than those of people who are not related . Moreover , the communities of microorganisms found in the intestines of non-related adults living in the same household are more similar than those of unrelated adults living in different households . We also know that the range of microorganisms found in the intestine changes dramatically between birth and the age of 3 years . However , these studies have focused on the intestine , and little is known about the effect of relatedness , cohabitation and age on the microbiota at other body sites . Song et al . compared the microorganisms found on the skin , on the tongue and in the intestines of 159 people—and 36 dogs—in 60 families . They found that co-habitation resulted in the communities of microorganisms being more similar to each other , with those on the skin being the most similar . This was true for all comparisons , including human pairs , dog pairs and human–dog pairs . This suggests that humans probably acquire many of the microorganisms on their skin through direct contact with their surroundings , and that humans tend to share more microbes with individuals , including their pets , with which they are in frequent contact . Song et al . also discovered that , unlike what happens in the intestine , the microbial communities on the skin and tongue of infants and children were relatively similar to those of adults . Overall , these findings suggest that the communities of microorganisms found in the intestine changes with age in a way that differs significantly from those found on the skin and tongue .'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "json.loads(elife_test[\"3shot\"][3])[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'elife_train_val_abstracts' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[16], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43melife_train_val_abstracts\u001b[49m[np\u001b[38;5;241m.\u001b[39margmax(similarity, \u001b[38;5;241m1\u001b[39m)][\u001b[38;5;241m3\u001b[39m]\n",
      "\u001b[0;31mNameError\u001b[0m: name 'elife_train_val_abstracts' is not defined"
     ]
    }
   ],
   "source": [
    "elife_train_val_abstracts[np.argmax(similarity, 1)][3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_top_shot_prompt(df, i):\n",
    "\n",
    "    topshot = json.loads(df[\"3shot\"][i])[0]\n",
    "    return f\"Example: {topshot}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Example: {'Abstract': 'The integration of most membrane proteins into the cytoplasmic membrane of bacteria occurs co-translationally . The universally conserved YidC protein mediates this process either individually as a membrane protein insertase , or in concert with the SecY complex . Here , we present a structural model of YidC based on evolutionary co-variation analysis , lipid-versus-protein-exposure and molecular dynamics simulations . The model suggests a distinctive arrangement of the conserved five transmembrane domains and a helical hairpin between transmembrane segment 2 ( TM2 ) and TM3 on the cytoplasmic membrane surface . The model was used for docking into a cryo-electron microscopy reconstruction of a translating YidC-ribosome complex carrying the YidC substrate FOc . This structure reveals how a single copy of YidC interacts with the ribosome at the ribosomal tunnel exit and identifies a site for membrane protein insertion at the YidC protein-lipid interface . Together , these data suggest a mechanism for the co-translational mode of YidC-mediated membrane protein insertion', 'Lay Summary': 'Cells are surrounded by a plasma membrane that acts like a barrier to help to keep the cell intact . Proteins are embedded in this plasma membrane; and some of these membrane proteins act as channels that allow molecules to enter and leave the cell , while others allow the cell to communicate with its surroundings . Like all proteins , membrane proteins are chains of amino acids that are joined together by a molecular machine called a ribosome . Most membrane proteins are inserted into the membrane as they are being built . All bacteria contain a protein called YidC that inserts proteins into the plasma membrane of bacterial cells . However , the mechanism behind this activity and the parts of the YidC protein that interact with the ribosome and plasma membrane are unknown . Wickles et al . have now used data from a range of sources to predict the three-dimensional structure of the YidC protein taken from a bacterium called E . coli . The model shows how the YidC protein is threaded back-and-forth through the membrane , a total of five times . Some of the protein also extends into the inside of the bacterial cell . Wickles et al . then used a technique called cyro-electron microscopy to look at the structure of a YidC protein bound to a ribosome that is building a new protein . Fitting the more detailed model of YidC into this overall structure of the whole complex revealed how a single YidC protein might interact with the ribosome to insert a newly built protein into a membrane . Wickles et al . then used a combination of theoretical modeling and other experiments to identify the amino acids in the YidC protein that bind to the ribosome: as expected , the binding takes place where the newly formed protein chain exits the ribosome . Further experiments also identified the amino acids in the YidC protein that interact with the newly built membrane protein , thus revealing where it might leave the YidC protein and be inserted into the membrane . The next challenge will be to investigate how the YidC protein assists the folding of new membrane proteins into their own highly specific three-dimensional structure .'}\""
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "make_top_shot_prompt(elife_test, 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAh8AAAGzCAYAAACPa3XZAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8fJSN1AAAACXBIWXMAAA9hAAAPYQGoP6dpAAA0a0lEQVR4nO3de3hU1d328XtCJpMTSUgCCYEQUJSzUMMpooIQiBQRSyqKWJHioS1YJYqWqhBQC2KrqA0gFoM+j6jFitSqYEQEWwISkCpKI1YQCyRUIAnHYSDr/cM38zgkJAyZrEDy/VzXXLjXXnvvtX+zZW72YcZhjDECAACwJKi+BwAAABoXwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHcA4ZMGCAunbtamVbTzzxhC644AI1adJEPXr0sLLNquzYsUMOh0O///3v620MAOwifKDBWbRokRwOhwoKCup7KFXavXu3srOztXnz5nobw3vvvaf7779f/fr1U25urn73u9+dtu/SpUuVkZGhpKQkuVwutW7dWj/96U+1ZcsWiyP2z9q1a5Wdna2SkpL6Hkq9mjt3rhYtWlTfwwAqCa7vAQCNze7duzV9+nS1bdu23s44fPDBBwoKCtLChQsVEhJSbd/PPvtMzZo109133634+HgVFRXphRdeUO/evZWfn6/u3btbGvWZW7t2raZPn65bb71VMTEx9T2cejN37lzFx8fr1ltvre+hAD4IH0AjtHfvXoWFhdUYPCRp6tSpldpuu+02tW7dWvPmzdP8+fPrYojnjWPHjikkJERBQZxIBs4U/7eg0dq1a5d+/vOfKyEhQS6XS126dNELL7zg0+fDDz+Uw+HQn//8Zz322GNq3bq1QkNDNWjQIH311VeV1pmTk6MLLrhAYWFh6t27tz766CMNGDBAAwYM8K6vV69ekqRx48bJ4XDI4XBUOjX+xRdf6KqrrlJ4eLhatWql2bNnn9E+nThxQo888oguvPBCuVwutW3bVr/97W/ldru9fRwOh3Jzc3X48OHTbr8mLVq0UHh4+Bld1igoKFBGRobi4+MVFhamdu3a6ec//3mVfRcsWOAde69evbRhw4ZKfT744ANdccUVioiIUExMjEaMGKGtW7d652dnZ2vy5MmSpHbt2nn3cceOHacdY8W9Nhs3btRll13mHeepwarieHj11Vf10EMPqVWrVgoPD1dZWZkkacmSJUpNTVVYWJji4+N18803a9euXT7ruPXWWxUZGamdO3fqmmuuUWRkpFq1aqWcnBxJ359pGjhwoCIiIpSSkqLFixf7LF9xWXHNmjW68847FRcXp6ioKN1yyy06cOCAt1/btm31+eefa/Xq1d4aVByHHo9H06dP10UXXaTQ0FDFxcXp8ssvV15e3mlrBAQSZz7QKBUXF6tv375yOByaOHGimjdvrnfffVfjx49XWVmZ7rnnHp/+s2bNUlBQkO677z6VlpZq9uzZGjNmjNavX+/tM2/ePE2cOFFXXHGFJk2apB07dui6665Ts2bN1Lp1a0lSp06dNGPGDE2dOlV33HGHrrjiCknSZZdd5l3PgQMHdPXVV2vkyJEaNWqUXn/9dT3wwAPq1q2bhg4dWu1+3XbbbXrxxRf105/+VPfee6/Wr1+vmTNnauvWrVq6dKkk6X/+53+0YMECffzxx/rTn/5UafunU1JSIo/Ho6KiIs2ZM0dlZWUaNGhQtcvs3btXQ4YMUfPmzfWb3/xGMTEx2rFjh954441KfRcvXqyDBw/qzjvvlMPh0OzZszVy5Eh9/fXXcjqdkqT3339fQ4cO1QUXXKDs7GwdPXpUzz77rPr166dNmzapbdu2GjlypL788ku98soreuqppxQfHy9Jat68ebVjPXDggH784x9r1KhRGj16tP785z/rl7/8pUJCQiqFpUceeUQhISG677775Ha7FRISokWLFmncuHHq1auXZs6cqeLiYj399NP6xz/+oU8++cTn8s/Jkyc1dOhQXXnllZo9e7ZefvllTZw4UREREXrwwQc1ZswYjRw5UvPnz9ctt9yitLQ0tWvXzmcMEydOVExMjLKzs1VYWKh58+bpm2++8QakOXPm6K677lJkZKQefPBBSVJCQoKk7wPazJkzddttt6l3794qKytTQUGBNm3apMGDB1dbJyAgDNDA5ObmGklmw4YNp+0zfvx407JlS/Pdd9/5tN94440mOjraHDlyxBhjzKpVq4wk06lTJ+N2u739nn76aSPJfPbZZ8YYY9xut4mLizO9evUyHo/H22/RokVGkunfv7+3bcOGDUaSyc3NrTSu/v37G0nmpZde8ra53W6TmJhoMjMzq93vzZs3G0nmtttu82m/7777jCTzwQcfeNvGjh1rIiIiql3fqTp06GAkGUkmMjLSPPTQQ+bkyZPVLrN06dIa34vt27cbSSYuLs7s37/f275s2TIjybz11lveth49epgWLVqYffv2edv++c9/mqCgIHPLLbd425544gkjyWzfvv2M9q2i7n/4wx+8bW6327u948ePG2P+73i44IILvMeIMcYcP37ctGjRwnTt2tUcPXrU2/63v/3NSDJTp071to0dO9ZIMr/73e+8bQcOHDBhYWHG4XCYV1991dv+r3/9y0gy06ZN87ZVHN+pqanecRljzOzZs40ks2zZMm9bly5dfI69Ct27dzfDhg07o9oAdYHLLmh0jDH6y1/+ouHDh8sYo++++877ysjIUGlpqTZt2uSzzLhx43zuj6g4Y/H1119L+v7Swr59+3T77bcrOPj/TiiOGTNGzZo182t8kZGRuvnmm73TISEh6t27t3dbp/POO+9IkrKysnza7733XknS22+/7dc4TpWbm6vly5dr7ty56tSpk44ePaqTJ09Wu0zFv/b/9re/yePxVNv3hhtu8KnVqTXes2ePNm/erFtvvVWxsbHefpdccokGDx7s3f+zFRwcrDvvvNM7HRISojvvvFN79+7Vxo0bffqOHTtWYWFh3umCggLt3btXv/rVrxQaGuptHzZsmDp27Fhl7W+77Tbvf8fExKhDhw6KiIjQqFGjvO0dOnRQTExMle/9HXfc4T0jJEm//OUvFRwcfEZ1iImJ0eeff65t27bV2BeoC4QPNDr//e9/VVJSogULFqh58+Y+r3Hjxkn6/nLBD7Vp08ZnuuJDsuIa+zfffCNJat++vU+/4OBgtW3b1q/xtW7dWg6Ho9L2fng9vyrffPONgoKCKo0hMTFRMTEx3jGerbS0NGVkZOiXv/ylVqxYof/93//VlClTql2mf//+yszM1PTp0xUfH68RI0YoNzfX5x6UCmda4w4dOlRatlOnTvruu+90+PDhs9o3SUpKSlJERIRP28UXXyxJle4XOfUSSHVj69ixY6Xah4aGVroMFB0dXeV7Hx0dXeV7f9FFF/lMR0ZGqmXLltXe21JhxowZKikp0cUXX6xu3bpp8uTJ+vTTT2tcDggUwgcanfLycknSzTffrLy8vCpf/fr181mmSZMmVa7LGBPw8dV2W6d+eNWFZs2aaeDAgXr55ZdrHMvrr7+u/Px8TZw40XuTb2pqqg4dOuTT12aNa+uHZz3Oxun21VYNrrzySv373//WCy+8oK5du+pPf/qTLr30Uu89QEBdI3yg0WnevLmaNm2qkydPKj09vcpXixYt/FpnSkqKJFV6AubEiROV/iVaV+EgJSVF5eXllU6lFxcXq6SkxDvGQDl69KhKS0vPqG/fvn312GOPqaCgQC+//LI+//xzvfrqq35tr2L8hYWFleb961//Unx8vPfMxdnUePfu3ZXOnHz55ZeSVOPZq+rGVlhYGPDaS6r0Ph86dEh79uzxGWt1dYiNjdW4ceP0yiuv6Ntvv9Ull1yi7OzsgI8TqArhA41OkyZNlJmZqb/85S9Vfkvnf//7X7/X2bNnT8XFxen555/XiRMnvO0vv/xypVPmFR+Qgf72zR//+MeSpDlz5vi0P/nkk5K+v//gbJx6CUr6/jLEypUr1bNnz2qXPXDgQKV/tVd8sVpVl16q07JlS/Xo0UMvvviiT+22bNmi9957z7v/0tnV+MSJE3ruuee808ePH9dzzz2n5s2bKzU1tdple/bsqRYtWmj+/Pk++/Xuu+9q69atZ1376ixYsMDnPpp58+bpxIkTPk9ERUREVFmDffv2+UxHRkaqffv2fr8nwNniUVs0WC+88IKWL19eqf3uu+/WrFmztGrVKvXp00e33367OnfurP3792vTpk16//33tX//fr+2FRISouzsbN11110aOHCgRo0apR07dmjRokW68MILff4FeuGFFyomJkbz589X06ZNFRERoT59+lS6j8Bf3bt319ixY7VgwQKVlJSof//++vjjj/Xiiy/quuuu01VXXXVW6+3WrZsGDRqkHj16qFmzZtq2bZsWLlwoj8ejWbNmVbvsiy++qLlz5+onP/mJLrzwQh08eFDPP/+8oqKifMLCmXriiSc0dOhQpaWlafz48d5HbaOjo33+1V4RFh588EHdeOONcjqdGj58eKV7On4oKSlJjz/+uHbs2KGLL75Yr732mjZv3qwFCxb43NhZFafTqccff1zjxo1T//79NXr0aO+jtm3bttWkSZP83teaHD9+XIMGDdKoUaNUWFiouXPn6vLLL9e1117r7ZOamqp58+bp0UcfVfv27dWiRQsNHDhQnTt31oABA5SamqrY2FgVFBTo9ddf18SJEwM+TqBK9fikDVAnKh5FPN3r22+/NcYYU1xcbCZMmGCSk5ON0+k0iYmJZtCgQWbBggXedVU8WrlkyRKfbVQ8Hnrq47LPPPOMSUlJMS6Xy/Tu3dv84x//MKmpqebqq6/26bds2TLTuXNnExwc7LOe/v37my5dulTap7Fjx5qUlJQa993j8Zjp06ebdu3aGafTaZKTk82UKVPMsWPHKq3vTB+1nTZtmunZs6dp1qyZCQ4ONklJSebGG280n376aY3Lbtq0yYwePdq0adPGuFwu06JFC3PNNdeYgoICb5+KWj7xxBOVltcpj5kaY8z7779v+vXrZ8LCwkxUVJQZPny4+eKLLyot+8gjj5hWrVqZoKCgGh+7rah7QUGBSUtLM6GhoSYlJcX88Y9/9Ol3uuOhwmuvvWZ+9KMfGZfLZWJjY82YMWPMf/7zH58+p6v96d77lJQUn8diK47v1atXmzvuuMM0a9bMREZGmjFjxvg8gmyMMUVFRWbYsGGmadOmPo98P/roo6Z3794mJibGhIWFmY4dO5rHHnvM59FdoC45jDkH7+YCGojy8nI1b95cI0eO1PPPP1/fw8FpDBgwQN999905/WN5FSq+zGzDhg01XvYCzlXc8wEEyLFjxyrd3/DSSy9p//793q+1BgBwzwcQMOvWrdOkSZN0/fXXKy4uTps2bdLChQvVtWtXXX/99fU9PAA4ZxA+gABp27atkpOT9cwzz2j//v2KjY3VLbfcolmzZp3Rr8cCQGPBPR8AAMAq7vkAAABWET4AAIBV59w9H+Xl5dq9e7eaNm1q5TcqAABA7RljdPDgQSUlJSkoqPpzG+dc+Ni9e7eSk5PrexgAAOAsfPvtt2rdunW1fc658NG0aVNJ3w8+KiqqnkfTMHg8Hr333nsaMmRIjV8TjbNHne2gznWPGtvR0OpcVlam5ORk7+d4dc658FFxqSUqKorwESAej0fh4eGKiopqEAf4uYo620Gd6x41tqOh1vlMbpnghlMAAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFgVXN8DQM3a/ubtWi3vamI0u7fUNXuF3Cdr/qnjQNgxa5iV7QAAzj+c+QAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFjld/jYtWuXbr75ZsXFxSksLEzdunVTQUGBd74xRlOnTlXLli0VFham9PR0bdu2LaCDBgAA5y+/wseBAwfUr18/OZ1Ovfvuu/riiy/0hz/8Qc2aNfP2mT17tp555hnNnz9f69evV0REhDIyMnTs2LGADx4AAJx/gv3p/Pjjjys5OVm5ubnetnbt2nn/2xijOXPm6KGHHtKIESMkSS+99JISEhL05ptv6sYbbwzQsAEAwPnKr/Dx17/+VRkZGbr++uu1evVqtWrVSr/61a90++23S5K2b9+uoqIipaene5eJjo5Wnz59lJ+fX2X4cLvdcrvd3umysjJJksfjkcfjOaudamhcTUztlg8yPn/a0Bjfu4p9boz7bhN1rnvU2I6GVmd/9sNhjDnjT6TQ0FBJUlZWlq6//npt2LBBd999t+bPn6+xY8dq7dq16tevn3bv3q2WLVt6lxs1apQcDodee+21SuvMzs7W9OnTK7UvXrxY4eHhZ7wjAACg/hw5ckQ33XSTSktLFRUVVW1fv8JHSEiIevbsqbVr13rbfv3rX2vDhg3Kz88/q/BR1ZmP5ORkfffddzUOvrHomr2iVsu7gowe6VmuhwuC5C53BGhU1duSnWFlO+cSj8ejvLw8DR48WE6ns76H02BR57pHje1oaHUuKytTfHz8GYUPvy67tGzZUp07d/Zp69Spk/7yl79IkhITEyVJxcXFPuGjuLhYPXr0qHKdLpdLLperUrvT6WwQb0YguE8GJjC4yx0BW1dNGvN7x7FrB3Wue9TYjoZSZ3/2wa+nXfr166fCwkKfti+//FIpKSmSvr/5NDExUStXrvTOLysr0/r165WWlubPpgAAQAPl15mPSZMm6bLLLtPvfvc7jRo1Sh9//LEWLFigBQsWSJIcDofuuecePfroo7rooovUrl07Pfzww0pKStJ1111XF+MHAADnGb/CR69evbR06VJNmTJFM2bMULt27TRnzhyNGTPG2+f+++/X4cOHdccdd6ikpESXX365li9f7r1ZFQAANG5+hQ9Juuaaa3TNNdecdr7D4dCMGTM0Y8aMWg0MAAA0TPy2CwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALDKr/CRnZ0th8Ph8+rYsaN3/rFjxzRhwgTFxcUpMjJSmZmZKi4uDvigAQDA+cvvMx9dunTRnj17vK+///3v3nmTJk3SW2+9pSVLlmj16tXavXu3Ro4cGdABAwCA81uw3wsEBysxMbFSe2lpqRYuXKjFixdr4MCBkqTc3Fx16tRJ69atU9++fWs/WgAAcN7zO3xs27ZNSUlJCg0NVVpammbOnKk2bdpo48aN8ng8Sk9P9/bt2LGj2rRpo/z8/NOGD7fbLbfb7Z0uKyuTJHk8Hnk8Hn+H1yC5mpjaLR9kfP60oTG+dxX73Bj33SbqXPeosR0Nrc7+7IfDGHPGn0jvvvuuDh06pA4dOmjPnj2aPn26du3apS1btuitt97SuHHjfIKEJPXu3VtXXXWVHn/88SrXmZ2drenTp1dqX7x4scLDw894RwAAQP05cuSIbrrpJpWWlioqKqravn6Fj1OVlJQoJSVFTz75pMLCws4qfFR15iM5OVnfffddjYNvLLpmr6jV8q4go0d6luvhgiC5yx0BGlX1tmRnWNnOucTj8SgvL0+DBw+W0+ms7+E0WNS57lFjOxpancvKyhQfH39G4cPvyy4/FBMTo4svvlhfffWVBg8erOPHj6ukpEQxMTHePsXFxVXeI1LB5XLJ5XJVanc6nQ3izQgE98nABAZ3uSNg66pJY37vOHbtoM51jxrb0VDq7M8+1Op7Pg4dOqR///vfatmypVJTU+V0OrVy5Urv/MLCQu3cuVNpaWm12QwAAGhA/Drzcd9992n48OFKSUnR7t27NW3aNDVp0kSjR49WdHS0xo8fr6ysLMXGxioqKkp33XWX0tLSeNIFAAB4+RU+/vOf/2j06NHat2+fmjdvrssvv1zr1q1T8+bNJUlPPfWUgoKClJmZKbfbrYyMDM2dO7dOBg4AAM5PfoWPV199tdr5oaGhysnJUU5OTq0GBQAAGi5+2wUAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYVavwMWvWLDkcDt1zzz3etmPHjmnChAmKi4tTZGSkMjMzVVxcXNtxAgCABuKsw8eGDRv03HPP6ZJLLvFpnzRpkt566y0tWbJEq1ev1u7duzVy5MhaDxQAADQMZxU+Dh06pDFjxuj5559Xs2bNvO2lpaVauHChnnzySQ0cOFCpqanKzc3V2rVrtW7duoANGgAAnL+Cz2ahCRMmaNiwYUpPT9ejjz7qbd+4caM8Ho/S09O9bR07dlSbNm2Un5+vvn37VlqX2+2W2+32TpeVlUmSPB6PPB7P2QyvwXE1MbVbPsj4/GlDY3zvKva5Me67TdS57lFjOxpanf3ZD7/Dx6uvvqpNmzZpw4YNleYVFRUpJCREMTExPu0JCQkqKiqqcn0zZ87U9OnTK7W/9957Cg8P93d4DdLs3oFZzyM9ywOzojPwzjvvWNvWuSYvL6++h9AoUOe6R43taCh1PnLkyBn39St8fPvtt7r77ruVl5en0NBQvwdWlSlTpigrK8s7XVZWpuTkZA0ZMkRRUVEB2cb5rmv2ilot7woyeqRnuR4uCJK73BGgUVVvS3aGle2cSzwej/Ly8jR48GA5nc76Hk6DRZ3rHjW2o6HVueLKxZnwK3xs3LhRe/fu1aWXXuptO3nypNasWaM//vGPWrFihY4fP66SkhKfsx/FxcVKTEyscp0ul0sul6tSu9PpbBBvRiC4TwYmMLjLHQFbV00a83vHsWsHda571NiOhlJnf/bBr/AxaNAgffbZZz5t48aNU8eOHfXAAw8oOTlZTqdTK1euVGZmpiSpsLBQO3fuVFpamj+bAgAADZRf4aNp06bq2rWrT1tERITi4uK87ePHj1dWVpZiY2MVFRWlu+66S2lpaVXebAoAABqfs3rapTpPPfWUgoKClJmZKbfbrYyMDM2dOzfQmwEAAOepWoePDz/80Gc6NDRUOTk5ysnJqe2qAQBAA8RvuwAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMCqgP+2CyBJbX/zdn0PwW87Zg2r7yEAQKPAmQ8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABY5Vf4mDdvni655BJFRUUpKipKaWlpevfdd73zjx07pgkTJiguLk6RkZHKzMxUcXFxwAcNAADOX36Fj9atW2vWrFnauHGjCgoKNHDgQI0YMUKff/65JGnSpEl66623tGTJEq1evVq7d+/WyJEj62TgAADg/BTsT+fhw4f7TD/22GOaN2+e1q1bp9atW2vhwoVavHixBg4cKEnKzc1Vp06dtG7dOvXt2zdwowYAAOctv8LHD508eVJLlizR4cOHlZaWpo0bN8rj8Sg9Pd3bp2PHjmrTpo3y8/NPGz7cbrfcbrd3uqysTJLk8Xjk8XjOdngNiquJqd3yQcbnT1SttsdbxfIct3WLOtc9amxHQ6uzP/vhd/j47LPPlJaWpmPHjikyMlJLly5V586dtXnzZoWEhCgmJsanf0JCgoqKik67vpkzZ2r69OmV2t977z2Fh4f7O7wGaXbvwKznkZ7lgVlRA/XOO+8EZD15eXkBWQ+qR53rHjW2o6HU+ciRI2fc1+/w0aFDB23evFmlpaV6/fXXNXbsWK1evdrf1XhNmTJFWVlZ3umysjIlJydryJAhioqKOuv1NiRds1fUanlXkNEjPcv1cEGQ3OWOAI2q4dmSnVGr5T0ej/Ly8jR48GA5nc4AjQqnos51jxrb0dDqXHHl4kz4HT5CQkLUvn17SVJqaqo2bNigp59+WjfccIOOHz+ukpISn7MfxcXFSkxMPO36XC6XXC5XpXan09kg3oxAcJ8MTGBwlzsCtq6GKFDHG8euHdS57lFjOxpKnf3Zh1p/z0d5ebncbrdSU1PldDq1cuVK77zCwkLt3LlTaWlptd0MAABoIPw68zFlyhQNHTpUbdq00cGDB7V48WJ9+OGHWrFihaKjozV+/HhlZWUpNjZWUVFRuuuuu5SWlsaTLgAAwMuv8LF3717dcsst2rNnj6Kjo3XJJZdoxYoVGjx4sCTpqaeeUlBQkDIzM+V2u5WRkaG5c+fWycDPVtvfvF3fQwAAoFHzK3wsXLiw2vmhoaHKyclRTk5OrQYFAAAaLn7bBQAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFjlV/iYOXOmevXqpaZNm6pFixa67rrrVFhY6NPn2LFjmjBhguLi4hQZGanMzEwVFxcHdNAAAOD85Vf4WL16tSZMmKB169YpLy9PHo9HQ4YM0eHDh719Jk2apLfeektLlizR6tWrtXv3bo0cOTLgAwcAAOenYH86L1++3Gd60aJFatGihTZu3Kgrr7xSpaWlWrhwoRYvXqyBAwdKknJzc9WpUyetW7dOffv2DdzIAQDAecmv8HGq0tJSSVJsbKwkaePGjfJ4PEpPT/f26dixo9q0aaP8/Pwqw4fb7Zbb7fZOl5WVSZI8Ho88Hk9thlclVxMT8HWe61xBxudPVK22x1vF8nVx3OL/UOe6R43taGh19mc/HMaYs/pEKi8v17XXXquSkhL9/e9/lyQtXrxY48aN8wkTktS7d29dddVVevzxxyutJzs7W9OnT6/UvnjxYoWHh5/N0AAAgGVHjhzRTTfdpNLSUkVFRVXb96zPfEyYMEFbtmzxBo+zNWXKFGVlZXmny8rKlJycrCFDhtQ4+LPRNXtFwNd5rnMFGT3Ss1wPFwTJXe6o7+Gcs7ZkZ9RqeY/Ho7y8PA0ePFhOpzNAo8KpqHPdo8Z2NLQ6V1y5OBNnFT4mTpyov/3tb1qzZo1at27tbU9MTNTx48dVUlKimJgYb3txcbESExOrXJfL5ZLL5arU7nQ66+TNcJ9svB++7nJHo97/mgTqeKurYxe+qHPdo8Z2NJQ6+7MPfj3tYozRxIkTtXTpUn3wwQdq166dz/zU1FQ5nU6tXLnS21ZYWKidO3cqLS3Nn00BAIAGyq8zHxMmTNDixYu1bNkyNW3aVEVFRZKk6OhohYWFKTo6WuPHj1dWVpZiY2MVFRWlu+66S2lpaTzpAgAAJPkZPubNmydJGjBggE97bm6ubr31VknSU089paCgIGVmZsrtdisjI0Nz584NyGABAMD5z6/wcSYPxoSGhionJ0c5OTlnPSgAANBw8dsuAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCq/w8eaNWs0fPhwJSUlyeFw6M033/SZb4zR1KlT1bJlS4WFhSk9PV3btm0L1HgBAMB5zu/wcfjwYXXv3l05OTlVzp89e7aeeeYZzZ8/X+vXr1dERIQyMjJ07NixWg8WAACc/4L9XWDo0KEaOnRolfOMMZozZ44eeughjRgxQpL00ksvKSEhQW+++aZuvPHGSsu43W653W7vdFlZmSTJ4/HI4/H4O7wauZqYgK/zXOcKMj5/omq1Pd4qlq+L4xb/hzrXPWpsR0Orsz/74TDGnPUnksPh0NKlS3XddddJkr7++mtdeOGF+uSTT9SjRw9vv/79+6tHjx56+umnK60jOztb06dPr9S+ePFihYeHn+3QAACARUeOHNFNN92k0tJSRUVFVdvX7zMf1SkqKpIkJSQk+LQnJCR4551qypQpysrK8k6XlZUpOTlZQ4YMqXHwZ6Nr9oqAr/Nc5woyeqRnuR4uCJK73FHfwzlnbcnOqNXyHo9HeXl5Gjx4sJxOZ4BGhVNR57pHje1oaHWuuHJxJgIaPs6Gy+WSy+Wq1O50OuvkzXCfbLwfvu5yR6Pe/5oE6nirq2MXvqhz3aPGdjSUOvuzDwF91DYxMVGSVFxc7NNeXFzsnQcAABq3gIaPdu3aKTExUStXrvS2lZWVaf369UpLSwvkpgAAwHnK78suhw4d0ldffeWd3r59uzZv3qzY2Fi1adNG99xzjx599FFddNFFateunR5++GElJSV5b0oFAACNm9/ho6CgQFdddZV3uuJm0bFjx2rRokW6//77dfjwYd1xxx0qKSnR5ZdfruXLlys0NDRwowYAAOctv8PHgAEDVN3TuQ6HQzNmzNCMGTNqNTAAANAw8dsuAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwKri+BwCcK9r+5u1aLe9qYjS7t9Q1e4XcJx0BGlXDs2PWsPoeAoB6xpkPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAVT7sAAFALZ/ukXH0+IVffT51x5gMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYRfgAAABW8fXqAKw626+irlAfX0ld319F3ZjU9vjA+aHOznzk5OSobdu2Cg0NVZ8+ffTxxx/X1aYAAMB5pE7Cx2uvvaasrCxNmzZNmzZtUvfu3ZWRkaG9e/fWxeYAAMB5pE7Cx5NPPqnbb79d48aNU+fOnTV//nyFh4frhRdeqIvNAQCA80jA7/k4fvy4Nm7cqClTpnjbgoKClJ6ervz8/Er93W633G63d7q0tFSStH//fnk8nkAPT8EnDgd8nee64HKjI0fKFewJ0slyuz/b3JhQZzvqo8779u2zsp1zhcfj0ZEjR7Rv3z45nU6r225Mf0fX598ZdXFMHzx4UJJkjKm5swmwXbt2GUlm7dq1Pu2TJ082vXv3rtR/2rRpRhIvXrx48eLFqwG8vv322xqzQr0/7TJlyhRlZWV5p8vLy7V//37FxcXJ4eBfj4FQVlam5ORkffvtt4qKiqrv4TRY1NkO6lz3qLEdDa3OxhgdPHhQSUlJNfYNePiIj49XkyZNVFxc7NNeXFysxMTESv1dLpdcLpdPW0xMTKCHBUlRUVEN4gA/11FnO6hz3aPGdjSkOkdHR59Rv4DfcBoSEqLU1FStXLnS21ZeXq6VK1cqLS0t0JsDAADnmTq57JKVlaWxY8eqZ8+e6t27t+bMmaPDhw9r3LhxdbE5AABwHqmT8HHDDTfov//9r6ZOnaqioiL16NFDy5cvV0JCQl1sDjVwuVyaNm1apctbCCzqbAd1rnvU2I7GXGeHMWfyTAwAAEBg8MNyAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwsc5as2aNRo+fLiSkpLkcDj05ptv+sw3xmjq1Klq2bKlwsLClJ6erm3btvn02b9/v8aMGaOoqCjFxMRo/PjxOnTokE+fTz/9VFdccYVCQ0OVnJys2bNnVxrLkiVL1LFjR4WGhqpbt2565513Ar6/9aW6Ons8Hj3wwAPq1q2bIiIilJSUpFtuuUW7d+/2WQd1rl5Nx/IP/eIXv5DD4dCcOXN82qlxzc6kzlu3btW1116r6OhoRUREqFevXtq5c6d3/rFjxzRhwgTFxcUpMjJSmZmZlb6teufOnRo2bJjCw8PVokULTZ48WSdOnPDp8+GHH+rSSy+Vy+VS+/bttWjRorrY5XpRU50PHTqkiRMnqnXr1goLC/P+svsPUWcp4D8sh8B45513zIMPPmjeeOMNI8ksXbrUZ/6sWbNMdHS0efPNN80///lPc+2115p27dqZo0ePevtcffXVpnv37mbdunXmo48+Mu3btzejR4/2zi8tLTUJCQlmzJgxZsuWLeaVV14xYWFh5rnnnvP2+cc//mGaNGliZs+ebb744gvz0EMPGafTaT777LM6r4EN1dW5pKTEpKenm9dee83861//Mvn5+aZ3794mNTXVZx3UuXo1HcsV3njjDdO9e3eTlJRknnrqKZ951LhmNdX5q6++MrGxsWby5Mlm06ZN5quvvjLLli0zxcXF3j6/+MUvTHJyslm5cqUpKCgwffv2NZdddpl3/okTJ0zXrl1Nenq6+eSTT8w777xj4uPjzZQpU7x9vv76axMeHm6ysrLMF198YZ599lnTpEkTs3z58jqvgQ011fn22283F154oVm1apXZvn27ee6550yTJk3MsmXLvH2oszGEj/PAqQd4eXm5SUxMNE888YS3raSkxLhcLvPKK68YY4z54osvjCSzYcMGb593333XOBwOs2vXLmOMMXPnzjXNmjUzbrfb2+eBBx4wHTp08E6PGjXKDBs2zGc8ffr0MXfeeWdA9/FcUN0HY4WPP/7YSDLffPONMYY6++t0Nf7Pf/5jWrVqZbZs2WJSUlJ8wgc19l9Vdb7hhhvMzTfffNplSkpKjNPpNEuWLPG2bd261Ugy+fn5xpjvP3iDgoJMUVGRt8+8efNMVFSUt/b333+/6dKlS6VtZ2Rk1Ha3zjlV1blLly5mxowZPm2XXnqpefDBB40x1LkCl13OQ9u3b1dRUZHS09O9bdHR0erTp4/y8/MlSfn5+YqJiVHPnj29fdLT0xUUFKT169d7+1x55ZUKCQnx9snIyFBhYaEOHDjg7fPD7VT0qdhOY1NaWiqHw+H98UPqXHvl5eX62c9+psmTJ6tLly6V5lPj2isvL9fbb7+tiy++WBkZGWrRooX69Onjc8lg48aN8ng8PjXq2LGj2rRp4/P3Srdu3Xy+rTojI0NlZWX6/PPPvX0aa50l6bLLLtNf//pX7dq1S8YYrVq1Sl9++aWGDBkiiTpXIHych4qKiiSp0tfVJyQkeOcVFRWpRYsWPvODg4MVGxvr06eqdfxwG6frUzG/MTl27JgeeOABjR492vsLlNS59h5//HEFBwfr17/+dZXzqXHt7d27V4cOHdKsWbN09dVX67333tNPfvITjRw5UqtXr5b0fX1CQkIq/ar4qX+vnG2dy8rKdPTo0brYvXPKs88+q86dO6t169YKCQnR1VdfrZycHF155ZWSqHOFOvltF6Ch8Xg8GjVqlIwxmjdvXn0Pp8HYuHGjnn76aW3atEkOh6O+h9NglZeXS5JGjBihSZMmSZJ69OihtWvXav78+erfv399Dq9BefbZZ7Vu3Tr99a9/VUpKitasWaMJEyYoKSmp0pmKxowzH+ehxMRESap0d3RxcbF3XmJiovbu3esz/8SJE9q/f79Pn6rW8cNtnK5PxfzGoCJ4fPPNN8rLy/Oe9ZCoc2199NFH2rt3r9q0aaPg4GAFBwfrm2++0b333qu2bdtKosaBEB8fr+DgYHXu3NmnvVOnTt6nXRITE3X8+HGVlJT49Dn175WzrXNUVJTCwsICtk/noqNHj+q3v/2tnnzySQ0fPlyXXHKJJk6cqBtuuEG///3vJVHnCoSP81C7du2UmJiolStXetvKysq0fv16paWlSZLS0tJUUlKijRs3evt88MEHKi8vV58+fbx91qxZI4/H4+2Tl5enDh06qFmzZt4+P9xORZ+K7TR0FcFj27Ztev/99xUXF+cznzrXzs9+9jN9+umn2rx5s/eVlJSkyZMna8WKFZKocSCEhISoV69eKiws9Gn/8ssvlZKSIklKTU2V0+n0qVFhYaF27tzp8/fKZ5995hMGKwJ5RbBpzHX2eDzyeDwKCvL9aG3SpIn37BN1/v/q+45XVO3gwYPmk08+MZ988omRZJ588knzySefeJ+ymDVrlomJiTHLli0zn376qRkxYkSVj9r+6Ec/MuvXrzd///vfzUUXXeTzeGJJSYlJSEgwP/vZz8yWLVvMq6++asLDwys9nhgcHGx+//vfm61bt5pp06Y1qMcTq6vz8ePHzbXXXmtat25tNm/ebPbs2eN9/fCpCupcvZqO5VOd+rSLMdT4TNRU5zfeeMM4nU6zYMECs23bNu+jmR999JF3Hb/4xS9MmzZtzAcffGAKCgpMWlqaSUtL886veAR0yJAhZvPmzWb58uWmefPmVT4COnnyZLN161aTk5NzXj0CWpOa6ty/f3/TpUsXs2rVKvP111+b3NxcExoaaubOnetdB3XmUdtz1qpVq4ykSq+xY8caY75/3Pbhhx82CQkJxuVymUGDBpnCwkKfdezbt8+MHj3aREZGmqioKDNu3Dhz8OBBnz7//Oc/zeWXX25cLpdp1aqVmTVrVqWx/PnPfzYXX3yxCQkJMV26dDFvv/12ne23bdXVefv27VXOk2RWrVrlXQd1rl5Nx/Kpqgof1LhmZ1LnhQsXmvbt25vQ0FDTvXt38+abb/qs4+jRo+ZXv/qVadasmQkPDzc/+clPzJ49e3z67NixwwwdOtSEhYWZ+Ph4c++99xqPx1NpLD169DAhISHmggsuMLm5uXW129bVVOc9e/aYW2+91SQlJZnQ0FDToUMH84c//MGUl5d710GdjXEYY0zdnlsBAAD4P9zzAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwKr/B8ip6XUw4VI5AAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "elife_3shot.FewShotExamples.apply(len).hist()\n",
    "plt.title(\"Length of 3 shot prompts\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use FlagEmbedding to calculate pair-wise embedding \n",
    "elife_train = pd.read_json(\"/home/dachengma/biolaysumm2024_data/eLife_train.jsonl\", lines=True)\n",
    "elife_val = pd.read_json(\"/home/dachengma/biolaysumm2024_data/eLife_val.jsonl\", lines=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------using 8*GPUs----------\n"
     ]
    }
   ],
   "source": [
    "from FlagEmbedding import BGEM3FlagModel\n",
    "\n",
    "model = BGEM3FlagModel('/home/dachengma/bge-m3',  \n",
    "                       use_fp16=True) # Setting use_fp16 to True speeds up computation with a slight performance degradation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_abstract(article):\n",
    "    return article.split(\"\\n\")[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "elife_train[\"abstract\"] = elife_train[\"article\"].apply(get_abstract)\n",
    "elife_val[\"abstract\"] = elife_val[\"article\"].apply(get_abstract)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "elife_test_abstracts = elife_test.abstract.tolist()\n",
    "elife_train_val_abstracts = elife_train[\"abstract\"].tolist() + elife_val[\"abstract\"].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(142, 4587)"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(elife_test_abstracts), len(elife_train_val_abstracts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Inference Embeddings: 100%|██████████| 48/48 [00:30<00:00,  1.57it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0.3145 0.4753 0.4604 ... 0.4954 0.4734 0.4963]\n",
      " [0.374  0.4148 0.477  ... 0.4937 0.4802 0.4917]\n",
      " [0.3955 0.469  0.4138 ... 0.4668 0.3848 0.4468]\n",
      " ...\n",
      " [0.3362 0.5024 0.3904 ... 0.5312 0.535  0.4363]\n",
      " [0.4238 0.3892 0.4727 ... 0.3796 0.3853 0.3804]\n",
      " [0.4148 0.4502 0.47   ... 0.429  0.4346 0.4312]]\n"
     ]
    }
   ],
   "source": [
    "embeddings_1 = model.encode(elife_test_abstracts, \n",
    "                            batch_size=12, \n",
    "                            max_length=8192, # If you don't need such a long length, you can set a smaller value to speed up the encoding process.\n",
    "                            )['dense_vecs']\n",
    "embeddings_2 = model.encode(elife_train_val_abstracts)['dense_vecs']\n",
    "similarity = embeddings_1 @ embeddings_2.T\n",
    "print(similarity)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save similarity to pickle\n",
    "import pickle\n",
    "with open('elife_similarity.pkl', 'wb') as f:\n",
    "    pickle.dump(similarity, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1843, 3202, 3917, 2042, 2684, 3513, 1239, 3383,  338,  713, 3199,\n",
       "       2538, 2674,  773,  120, 2935, 1220, 4151, 4237, 3258, 4553, 3557,\n",
       "       2533, 2883, 3045, 4244, 2820,   21, 3067, 4041, 1326, 2944, 3157,\n",
       "       2653,  545,  594, 3984, 1992, 4326, 3363, 3654, 2615,  297, 1075,\n",
       "       1350, 3177, 1512,  981, 3427, 3523, 3999, 4225,  628, 2470, 1518,\n",
       "       1789,  182, 3295, 1962, 4308, 1104, 4222, 3605, 1239, 2827,  177,\n",
       "        989, 3168,  328, 2108, 2275, 1442, 3039, 3409, 1448, 3051, 1475,\n",
       "       3674, 3461, 3070, 3368, 3641, 2965, 4563, 2559,   90, 2386, 2325,\n",
       "       2338, 3698, 4009, 1442, 4374, 1184,  160, 2977, 3850, 4320,  818,\n",
       "        295,  640, 2926,  904,  104, 4553, 1327,  375, 3695, 2332, 2538,\n",
       "       1645, 3480, 3961, 3097, 1994,  936, 4428, 3692,  242,  554, 2055,\n",
       "        795, 3956, 3982, 1116, 3257, 1173, 1607, 2614,  716, 2700,  634,\n",
       "       3192, 4524, 3434, 2987, 2799, 2170,   93, 2575, 3250, 1432])"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.argmax(similarity, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "elife_train_val_abstracts = np.array(elife_train_val_abstracts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "elife_test_topshot = elife_train_val_abstracts[np.argmax(similarity, 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "elife_train_val_summs = np.array(elife_train[\"lay_summary\"].to_list() + elife_val[\"lay_summary\"].to_list())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "elife_test_topshot_summ = elife_train_val_summs[np.argmax(similarity, 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "142"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(elife_test_topshot_summ)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "# do the same for plos\n",
    "plos_train = pd.read_json(\"/home/dachengma/biolaysumm2024_data/PLOS_train.jsonl\", lines=True)\n",
    "plos_val = pd.read_json(\"/home/dachengma/biolaysumm2024_data/PLOS_val.jsonl\", lines=True)\n",
    "plos_train[\"abstract\"] = plos_train[\"article\"].apply(get_abstract)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "plos_val[\"abstract\"] = plos_val[\"article\"].apply(get_abstract)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "plos_test_abstracts = plos_test.abstract.tolist()\n",
    "plos_train_val_abstracts = plos_train[\"abstract\"].tolist() + plos_val[\"abstract\"].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(142, 26149)"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(plos_test_abstracts), len(plos_train_val_abstracts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Inference Embeddings: 100%|██████████| 273/273 [03:44<00:00,  1.22it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0.4563 0.4702 0.395  ... 0.517  0.4084 0.3796]\n",
      " [0.403  0.4138 0.4749 ... 0.6562 0.4116 0.3562]\n",
      " [0.426  0.4548 0.4507 ... 0.4858 0.5083 0.398 ]\n",
      " ...\n",
      " [0.3845 0.4868 0.4216 ... 0.523  0.3794 0.31  ]\n",
      " [0.3923 0.5347 0.5146 ... 0.572  0.4004 0.4102]\n",
      " [0.3699 0.4202 0.4443 ... 0.4653 0.4624 0.436 ]]\n"
     ]
    }
   ],
   "source": [
    "embeddings_1_plos = model.encode(plos_test_abstracts, \n",
    "                            batch_size=12, \n",
    "                            max_length=8192, # If you don't need such a long length, you can set a smaller value to speed up the encoding process.\n",
    "                            )['dense_vecs']\n",
    "embeddings_2_plos = model.encode(plos_train_val_abstracts)['dense_vecs']\n",
    "similarity_plos = embeddings_1_plos @ embeddings_2_plos.T\n",
    "print(similarity_plos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "with open('plos_similarity.pkl', 'wb') as f:\n",
    "    pickle.dump(similarity_plos, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(142, 26149)"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "similarity_plos.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "plos_train_val_abstracts = np.array(plos_train_val_abstracts)\n",
    "plos_test_topshot = plos_train_val_abstracts[np.argmax(similarity_plos, 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "plos_train_val_summs = np.array(plos_train[\"lay_summary\"].to_list() + plos_val[\"lay_summary\"].to_list())\n",
    "plos_test_topshot_summ = plos_train_val_summs[np.argmax(similarity_plos, 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save elife_test_topshot to pickle\n",
    "with open(\"elife_topshot.pickle\", \"wb\") as f:\n",
    "    pickle.dump(elife_test_topshot, f)\n",
    "\n",
    "# same for plos\n",
    "with open(\"plos_topshot.pickle\", \"wb\") as f:\n",
    "    pickle.dump(plos_test_topshot, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "elife_test[\"topshot\"] = elife_test_topshot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "elife_test[\"topshot_laysumm\"] = elife_test_topshot_summ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "plos_test[\"topshot_laysumm\"] = plos_test_topshot_summ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "del elife_test[\"3shot\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "elife_test.to_csv(\"elife_test_with_top_shot.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "plos_test.to_csv(\"plos_test_with_top_shot.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_prompt(df, i):\n",
    "    # keywords = df[\"keywords\"][i]\n",
    "    title = df[\"title\"][i]\n",
    "    abstract = df[\"abstract\"][i]\n",
    "    return f\"Title: {title}\\n\\nAbstract: {abstract}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>article</th>\n",
       "      <th>headings</th>\n",
       "      <th>keywords</th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Lung-resident ( LR ) mesenchymal stem and stro...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Results', 'Discu...</td>\n",
       "      <td>['immune system', 'medical conditions', 'molec...</td>\n",
       "      <td>journal.ppat.1009789</td>\n",
       "      <td>Pulmonary mesenchymal stem cells are engaged i...</td>\n",
       "      <td>Lung-resident ( LR ) mesenchymal stem and stro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Visceral leishmaniasis ( VL ) is endemic in So...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Methods', 'Resul...</td>\n",
       "      <td>['neonates', 'clinical laboratory sciences', '...</td>\n",
       "      <td>journal.pntd.0007992</td>\n",
       "      <td>Outcomes of visceral leishmaniasis in pregnanc...</td>\n",
       "      <td>Visceral leishmaniasis ( VL ) is endemic in So...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A high burden of Salmonella enterica subspecie...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Methods', 'Resul...</td>\n",
       "      <td>['pathogens', 'medical conditions', 'taxonomy'...</td>\n",
       "      <td>journal.pntd.0010704</td>\n",
       "      <td>Molecular characterization of circulating Salm...</td>\n",
       "      <td>A high burden of Salmonella enterica subspecie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Severe Acute Respiratory Syndrome Coronavirus-...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Results', 'Discu...</td>\n",
       "      <td>['pathogens', 'amniotes', 'medical conditions'...</td>\n",
       "      <td>journal.ppat.1010691</td>\n",
       "      <td>Potent universal beta-coronavirus therapeutic ...</td>\n",
       "      <td>Severe Acute Respiratory Syndrome Coronavirus-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Many fungal species utilize hydroxyderivatives...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Results and disc...</td>\n",
       "      <td>['taxonomy', 'proteins', 'chemistry', 'genetic...</td>\n",
       "      <td>journal.pgen.1009815</td>\n",
       "      <td>Transcriptome and proteome profiling reveals c...</td>\n",
       "      <td>Many fungal species utilize hydroxyderivatives...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>Herpesviruses are known to encode a number of ...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Results', 'Discu...</td>\n",
       "      <td>['microbial mutation', 'cell processes', 'amyl...</td>\n",
       "      <td>journal.ppat.1008473</td>\n",
       "      <td>Varicella zoster virus encodes a viral decoy R...</td>\n",
       "      <td>Herpesviruses are known to encode a number of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>Infectious disease risk is driven by three int...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Methods', 'Resul...</td>\n",
       "      <td>[]</td>\n",
       "      <td>journal.pntd.0009806</td>\n",
       "      <td>Exposure, hazard, and vulnerability all contri...</td>\n",
       "      <td>Infectious disease risk is driven by three int...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>Nigeria continues to experience ever increasin...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Results', 'Discu...</td>\n",
       "      <td>['immune system', 'amniotes', 'lassa virus', '...</td>\n",
       "      <td>journal.ppat.1009966</td>\n",
       "      <td>Differential pathogenesis of closely related 2...</td>\n",
       "      <td>Nigeria continues to experience ever increasin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>Schistosomes are blood dwelling parasites that...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Methods and mate...</td>\n",
       "      <td>['schistosoma mansoni', 'medical conditions', ...</td>\n",
       "      <td>journal.pntd.0010728</td>\n",
       "      <td>Hematological and Biochemical changes in Schis...</td>\n",
       "      <td>Schistosomes are blood dwelling parasites that...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>Antibodies targeting the hepatitis C virus ( H...</td>\n",
       "      <td>['Abstract', 'Introduction', 'Results', 'Discu...</td>\n",
       "      <td>['white blood cells', 'pathogens', 'flow cytom...</td>\n",
       "      <td>journal.ppat.1010179</td>\n",
       "      <td>B cell overexpression of FCRL5 and PD-1 is ass...</td>\n",
       "      <td>Antibodies targeting the hepatitis C virus ( H...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>142 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               article  \\\n",
       "0    Lung-resident ( LR ) mesenchymal stem and stro...   \n",
       "1    Visceral leishmaniasis ( VL ) is endemic in So...   \n",
       "2    A high burden of Salmonella enterica subspecie...   \n",
       "3    Severe Acute Respiratory Syndrome Coronavirus-...   \n",
       "4    Many fungal species utilize hydroxyderivatives...   \n",
       "..                                                 ...   \n",
       "137  Herpesviruses are known to encode a number of ...   \n",
       "138  Infectious disease risk is driven by three int...   \n",
       "139  Nigeria continues to experience ever increasin...   \n",
       "140  Schistosomes are blood dwelling parasites that...   \n",
       "141  Antibodies targeting the hepatitis C virus ( H...   \n",
       "\n",
       "                                              headings  \\\n",
       "0    ['Abstract', 'Introduction', 'Results', 'Discu...   \n",
       "1    ['Abstract', 'Introduction', 'Methods', 'Resul...   \n",
       "2    ['Abstract', 'Introduction', 'Methods', 'Resul...   \n",
       "3    ['Abstract', 'Introduction', 'Results', 'Discu...   \n",
       "4    ['Abstract', 'Introduction', 'Results and disc...   \n",
       "..                                                 ...   \n",
       "137  ['Abstract', 'Introduction', 'Results', 'Discu...   \n",
       "138  ['Abstract', 'Introduction', 'Methods', 'Resul...   \n",
       "139  ['Abstract', 'Introduction', 'Results', 'Discu...   \n",
       "140  ['Abstract', 'Introduction', 'Methods and mate...   \n",
       "141  ['Abstract', 'Introduction', 'Results', 'Discu...   \n",
       "\n",
       "                                              keywords                    id  \\\n",
       "0    ['immune system', 'medical conditions', 'molec...  journal.ppat.1009789   \n",
       "1    ['neonates', 'clinical laboratory sciences', '...  journal.pntd.0007992   \n",
       "2    ['pathogens', 'medical conditions', 'taxonomy'...  journal.pntd.0010704   \n",
       "3    ['pathogens', 'amniotes', 'medical conditions'...  journal.ppat.1010691   \n",
       "4    ['taxonomy', 'proteins', 'chemistry', 'genetic...  journal.pgen.1009815   \n",
       "..                                                 ...                   ...   \n",
       "137  ['microbial mutation', 'cell processes', 'amyl...  journal.ppat.1008473   \n",
       "138                                                 []  journal.pntd.0009806   \n",
       "139  ['immune system', 'amniotes', 'lassa virus', '...  journal.ppat.1009966   \n",
       "140  ['schistosoma mansoni', 'medical conditions', ...  journal.pntd.0010728   \n",
       "141  ['white blood cells', 'pathogens', 'flow cytom...  journal.ppat.1010179   \n",
       "\n",
       "                                                 title  \\\n",
       "0    Pulmonary mesenchymal stem cells are engaged i...   \n",
       "1    Outcomes of visceral leishmaniasis in pregnanc...   \n",
       "2    Molecular characterization of circulating Salm...   \n",
       "3    Potent universal beta-coronavirus therapeutic ...   \n",
       "4    Transcriptome and proteome profiling reveals c...   \n",
       "..                                                 ...   \n",
       "137  Varicella zoster virus encodes a viral decoy R...   \n",
       "138  Exposure, hazard, and vulnerability all contri...   \n",
       "139  Differential pathogenesis of closely related 2...   \n",
       "140  Hematological and Biochemical changes in Schis...   \n",
       "141  B cell overexpression of FCRL5 and PD-1 is ass...   \n",
       "\n",
       "                                              abstract  \n",
       "0    Lung-resident ( LR ) mesenchymal stem and stro...  \n",
       "1    Visceral leishmaniasis ( VL ) is endemic in So...  \n",
       "2    A high burden of Salmonella enterica subspecie...  \n",
       "3    Severe Acute Respiratory Syndrome Coronavirus-...  \n",
       "4    Many fungal species utilize hydroxyderivatives...  \n",
       "..                                                 ...  \n",
       "137  Herpesviruses are known to encode a number of ...  \n",
       "138  Infectious disease risk is driven by three int...  \n",
       "139  Nigeria continues to experience ever increasin...  \n",
       "140  Schistosomes are blood dwelling parasites that...  \n",
       "141  Antibodies targeting the hepatitis C virus ( H...  \n",
       "\n",
       "[142 rows x 6 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plos_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [],
   "source": [
    "plos_test[\"topshot\"] = plos_test_topshot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Respiratory syncytial virus ( RSV ) is the most common cause of infant hospitalizations and severe RSV infections are a significant risk factor for childhood asthma . The pathogenic mechanisms responsible for RSV induced immunopathophysiology remain elusive . Using an age-appropriate mouse model of RSV , we show that IL-33 plays a critical role in the immunopathogenesis of severe RSV , which is associated with higher group 2 innate lymphoid cells ( ILC2s ) specifically in neonates . Infection with RSV induced rapid IL-33 expression and an increase in ILC2 numbers in the lungs of neonatal mice; this was not observed in adult mice . Blocking IL-33 with antibodies or using an IL-33 receptor knockout mouse during infection was sufficient to inhibit RSV immunopathogenesis ( i . e . , airway hyperresponsiveness , Th2 inflammation , eosinophilia , and mucus hyperproduction ) ; whereas administration of IL-33 to adult mice during RSV infection was sufficient to induce RSV disease . Additionally , elevated IL-33 and IL-13 were observed in nasal aspirates from infants hospitalized with RSV; these cytokines declined during convalescence . In summary , IL-33 is necessary , either directly or indirectly , to induce ILC2s and the Th2 biased immunopathophysiology observed following neonatal RSV infection . This study provides a mechanism involving IL-33 and ILC2s in RSV mediated human asthma . '"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plos_test[\"topshot\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'IL-33 is responsible for the immunopathophysiological response observed following neonatal RSV infection in mice . Its presence in nasal aspirates of human infants with severe RSV and suggests its role in disease severity and asthma .'"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plos_test[\"topshot_laysumm\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [],
   "source": [
    "plos_test.to_csv(\"plos_test_with_topshot.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_laysumm(df, i):\n",
    "    title = df.iloc[i]['title']\n",
    "    abstract = df.iloc[i]['abstract']\n",
    "    topshot = df.iloc[i]['topshot']\n",
    "    topshot_summ = df.iloc[i][\"topshot_laysumm\"]\n",
    "    messages = [\n",
    "        {\"role\": \"system\", \"content\": f\"You are an expert in writing lay summaries for biomedical literature, \\\n",
    "         for example, here is the most similar example of abstract and paired lay summary that you should learn to follow: sample abstract: {topshot}, \\\n",
    "         sample lay summary {topshot_summ}, you should learn as much as can from this example, and then write lay summary for the abstract that is given to you. \\\n",
    "         write in readable way but maximize relevance and factuality. only respond the actual lay summary, here is the given title and abstract,\"},\n",
    "        {\"role\": \"user\", \"content\": f\"{title}\\n\\n {abstract}\"}\n",
    "    ]\n",
    "    prompt = pipeline.tokenizer.apply_chat_template(\n",
    "        messages, \n",
    "        tokenize=False, \n",
    "        add_generation_prompt=True\n",
    "    )\n",
    "\n",
    "    terminators = [\n",
    "        pipeline.tokenizer.eos_token_id,\n",
    "        pipeline.tokenizer.convert_tokens_to_ids(\"<|eot_id|>\")\n",
    "    ]\n",
    "\n",
    "    outputs = pipeline(\n",
    "        prompt,\n",
    "        max_new_tokens=2048,\n",
    "        eos_token_id=terminators,\n",
    "        do_sample=True,\n",
    "        temperature=0.0001,\n",
    "        top_p=0.9,\n",
    "        \n",
    "    )\n",
    "    response = outputs[0][\"generated_text\"][len(prompt):]\n",
    "    return response\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "elife_test=pd.read_csv(\"/home/dachengma/l3ft/elife_test_with_top_shot.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "!export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/142 [01:09<?, ?it/s]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m elife_responses \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m tqdm(\u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(elife_test))):\n\u001b[0;32m----> 5\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mget_laysumm\u001b[49m\u001b[43m(\u001b[49m\u001b[43melife_test\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mi\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      6\u001b[0m     elife_responses\u001b[38;5;241m.\u001b[39mappend(response)\n",
      "Cell \u001b[0;32mIn[2], line 24\u001b[0m, in \u001b[0;36mget_laysumm\u001b[0;34m(df, i)\u001b[0m\n\u001b[1;32m     13\u001b[0m prompt \u001b[38;5;241m=\u001b[39m pipeline\u001b[38;5;241m.\u001b[39mtokenizer\u001b[38;5;241m.\u001b[39mapply_chat_template(\n\u001b[1;32m     14\u001b[0m     messages, \n\u001b[1;32m     15\u001b[0m     tokenize\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, \n\u001b[1;32m     16\u001b[0m     add_generation_prompt\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m     17\u001b[0m )\n\u001b[1;32m     19\u001b[0m terminators \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m     20\u001b[0m     pipeline\u001b[38;5;241m.\u001b[39mtokenizer\u001b[38;5;241m.\u001b[39meos_token_id,\n\u001b[1;32m     21\u001b[0m     pipeline\u001b[38;5;241m.\u001b[39mtokenizer\u001b[38;5;241m.\u001b[39mconvert_tokens_to_ids(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m<|eot_id|>\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     22\u001b[0m ]\n\u001b[0;32m---> 24\u001b[0m outputs \u001b[38;5;241m=\u001b[39m \u001b[43mpipeline\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     25\u001b[0m \u001b[43m    \u001b[49m\u001b[43mprompt\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     26\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmax_new_tokens\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m2048\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     27\u001b[0m \u001b[43m    \u001b[49m\u001b[43meos_token_id\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mterminators\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     28\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdo_sample\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     29\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtemperature\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0.0001\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     30\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtop_p\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0.9\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     31\u001b[0m \u001b[43m    \u001b[49m\n\u001b[1;32m     32\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     33\u001b[0m response \u001b[38;5;241m=\u001b[39m outputs[\u001b[38;5;241m0\u001b[39m][\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgenerated_text\u001b[39m\u001b[38;5;124m\"\u001b[39m][\u001b[38;5;28mlen\u001b[39m(prompt):]\n\u001b[1;32m     34\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m response\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/transformers/pipelines/text_generation.py:240\u001b[0m, in \u001b[0;36mTextGenerationPipeline.__call__\u001b[0;34m(self, text_inputs, **kwargs)\u001b[0m\n\u001b[1;32m    238\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39m\u001b[38;5;21m__call__\u001b[39m(chats, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    239\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 240\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__call__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mtext_inputs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/transformers/pipelines/base.py:1242\u001b[0m, in \u001b[0;36mPipeline.__call__\u001b[0;34m(self, inputs, num_workers, batch_size, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1234\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mnext\u001b[39m(\n\u001b[1;32m   1235\u001b[0m         \u001b[38;5;28miter\u001b[39m(\n\u001b[1;32m   1236\u001b[0m             \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_iterator(\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1239\u001b[0m         )\n\u001b[1;32m   1240\u001b[0m     )\n\u001b[1;32m   1241\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1242\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_single\u001b[49m\u001b[43m(\u001b[49m\u001b[43minputs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpreprocess_params\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mforward_params\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpostprocess_params\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/transformers/pipelines/base.py:1249\u001b[0m, in \u001b[0;36mPipeline.run_single\u001b[0;34m(self, inputs, preprocess_params, forward_params, postprocess_params)\u001b[0m\n\u001b[1;32m   1247\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrun_single\u001b[39m(\u001b[38;5;28mself\u001b[39m, inputs, preprocess_params, forward_params, postprocess_params):\n\u001b[1;32m   1248\u001b[0m     model_inputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpreprocess(inputs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mpreprocess_params)\n\u001b[0;32m-> 1249\u001b[0m     model_outputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mforward\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel_inputs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mforward_params\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1250\u001b[0m     outputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpostprocess(model_outputs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mpostprocess_params)\n\u001b[1;32m   1251\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m outputs\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/transformers/pipelines/base.py:1149\u001b[0m, in \u001b[0;36mPipeline.forward\u001b[0;34m(self, model_inputs, **forward_params)\u001b[0m\n\u001b[1;32m   1147\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m inference_context():\n\u001b[1;32m   1148\u001b[0m         model_inputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_ensure_tensor_on_device(model_inputs, device\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdevice)\n\u001b[0;32m-> 1149\u001b[0m         model_outputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_forward\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel_inputs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mforward_params\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1150\u001b[0m         model_outputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_ensure_tensor_on_device(model_outputs, device\u001b[38;5;241m=\u001b[39mtorch\u001b[38;5;241m.\u001b[39mdevice(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcpu\u001b[39m\u001b[38;5;124m\"\u001b[39m))\n\u001b[1;32m   1151\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/transformers/pipelines/text_generation.py:327\u001b[0m, in \u001b[0;36mTextGenerationPipeline._forward\u001b[0;34m(self, model_inputs, **generate_kwargs)\u001b[0m\n\u001b[1;32m    324\u001b[0m         generate_kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmin_length\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m prefix_length\n\u001b[1;32m    326\u001b[0m \u001b[38;5;66;03m# BS x SL\u001b[39;00m\n\u001b[0;32m--> 327\u001b[0m generated_sequence \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenerate\u001b[49m\u001b[43m(\u001b[49m\u001b[43minput_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minput_ids\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mattention_mask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mattention_mask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mgenerate_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    328\u001b[0m out_b \u001b[38;5;241m=\u001b[39m generated_sequence\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m    329\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mframework \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpt\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/torch/utils/_contextlib.py:115\u001b[0m, in \u001b[0;36mcontext_decorator.<locals>.decorate_context\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    112\u001b[0m \u001b[38;5;129m@functools\u001b[39m\u001b[38;5;241m.\u001b[39mwraps(func)\n\u001b[1;32m    113\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mdecorate_context\u001b[39m(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    114\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m ctx_factory():\n\u001b[0;32m--> 115\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/transformers/generation/utils.py:1622\u001b[0m, in \u001b[0;36mGenerationMixin.generate\u001b[0;34m(self, inputs, generation_config, logits_processor, stopping_criteria, prefix_allowed_tokens_fn, synced_gpus, assistant_model, streamer, negative_prompt_ids, negative_prompt_attention_mask, **kwargs)\u001b[0m\n\u001b[1;32m   1614\u001b[0m     input_ids, model_kwargs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_expand_inputs_for_generation(\n\u001b[1;32m   1615\u001b[0m         input_ids\u001b[38;5;241m=\u001b[39minput_ids,\n\u001b[1;32m   1616\u001b[0m         expand_size\u001b[38;5;241m=\u001b[39mgeneration_config\u001b[38;5;241m.\u001b[39mnum_return_sequences,\n\u001b[1;32m   1617\u001b[0m         is_encoder_decoder\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconfig\u001b[38;5;241m.\u001b[39mis_encoder_decoder,\n\u001b[1;32m   1618\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mmodel_kwargs,\n\u001b[1;32m   1619\u001b[0m     )\n\u001b[1;32m   1621\u001b[0m     \u001b[38;5;66;03m# 13. run sample\u001b[39;00m\n\u001b[0;32m-> 1622\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sample\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1623\u001b[0m \u001b[43m        \u001b[49m\u001b[43minput_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1624\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlogits_processor\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mprepared_logits_processor\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1625\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlogits_warper\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlogits_warper\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1626\u001b[0m \u001b[43m        \u001b[49m\u001b[43mstopping_criteria\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mprepared_stopping_criteria\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1627\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpad_token_id\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgeneration_config\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpad_token_id\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1628\u001b[0m \u001b[43m        \u001b[49m\u001b[43moutput_scores\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgeneration_config\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moutput_scores\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1629\u001b[0m \u001b[43m        \u001b[49m\u001b[43moutput_logits\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgeneration_config\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moutput_logits\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1630\u001b[0m \u001b[43m        \u001b[49m\u001b[43mreturn_dict_in_generate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgeneration_config\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreturn_dict_in_generate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1631\u001b[0m \u001b[43m        \u001b[49m\u001b[43msynced_gpus\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msynced_gpus\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1632\u001b[0m \u001b[43m        \u001b[49m\u001b[43mstreamer\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstreamer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1633\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mmodel_kwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1634\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1636\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m generation_mode \u001b[38;5;241m==\u001b[39m GenerationMode\u001b[38;5;241m.\u001b[39mBEAM_SEARCH:\n\u001b[1;32m   1637\u001b[0m     \u001b[38;5;66;03m# 11. prepare beam search scorer\u001b[39;00m\n\u001b[1;32m   1638\u001b[0m     beam_scorer \u001b[38;5;241m=\u001b[39m BeamSearchScorer(\n\u001b[1;32m   1639\u001b[0m         batch_size\u001b[38;5;241m=\u001b[39mbatch_size,\n\u001b[1;32m   1640\u001b[0m         num_beams\u001b[38;5;241m=\u001b[39mgeneration_config\u001b[38;5;241m.\u001b[39mnum_beams,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1645\u001b[0m         max_length\u001b[38;5;241m=\u001b[39mgeneration_config\u001b[38;5;241m.\u001b[39mmax_length,\n\u001b[1;32m   1646\u001b[0m     )\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/transformers/generation/utils.py:2791\u001b[0m, in \u001b[0;36mGenerationMixin._sample\u001b[0;34m(self, input_ids, logits_processor, stopping_criteria, logits_warper, max_length, pad_token_id, eos_token_id, output_attentions, output_hidden_states, output_scores, output_logits, return_dict_in_generate, synced_gpus, streamer, **model_kwargs)\u001b[0m\n\u001b[1;32m   2788\u001b[0m model_inputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprepare_inputs_for_generation(input_ids, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mmodel_kwargs)\n\u001b[1;32m   2790\u001b[0m \u001b[38;5;66;03m# forward pass to get next token\u001b[39;00m\n\u001b[0;32m-> 2791\u001b[0m outputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m   2792\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mmodel_inputs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2793\u001b[0m \u001b[43m    \u001b[49m\u001b[43mreturn_dict\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m   2794\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_attentions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_attentions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2795\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_hidden_states\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_hidden_states\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2796\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2798\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m synced_gpus \u001b[38;5;129;01mand\u001b[39;00m this_peer_finished:\n\u001b[1;32m   2799\u001b[0m     \u001b[38;5;28;01mcontinue\u001b[39;00m  \u001b[38;5;66;03m# don't waste resources running the code we don't need\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/torch/nn/modules/module.py:1532\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1530\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1531\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1532\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/torch/nn/modules/module.py:1541\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1536\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1537\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1538\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1539\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1540\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1541\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1543\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1544\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/transformers/models/llama/modeling_llama.py:1208\u001b[0m, in \u001b[0;36mLlamaForCausalLM.forward\u001b[0;34m(self, input_ids, attention_mask, position_ids, past_key_values, inputs_embeds, labels, use_cache, output_attentions, output_hidden_states, return_dict, cache_position)\u001b[0m\n\u001b[1;32m   1205\u001b[0m return_dict \u001b[38;5;241m=\u001b[39m return_dict \u001b[38;5;28;01mif\u001b[39;00m return_dict \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconfig\u001b[38;5;241m.\u001b[39muse_return_dict\n\u001b[1;32m   1207\u001b[0m \u001b[38;5;66;03m# decoder outputs consists of (dec_features, layer_state, dec_hidden, dec_attn)\u001b[39;00m\n\u001b[0;32m-> 1208\u001b[0m outputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmodel\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1209\u001b[0m \u001b[43m    \u001b[49m\u001b[43minput_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minput_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1210\u001b[0m \u001b[43m    \u001b[49m\u001b[43mattention_mask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mattention_mask\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1211\u001b[0m \u001b[43m    \u001b[49m\u001b[43mposition_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mposition_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1212\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpast_key_values\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpast_key_values\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1213\u001b[0m \u001b[43m    \u001b[49m\u001b[43minputs_embeds\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minputs_embeds\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1214\u001b[0m \u001b[43m    \u001b[49m\u001b[43muse_cache\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43muse_cache\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1215\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_attentions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_attentions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1216\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_hidden_states\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_hidden_states\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1217\u001b[0m \u001b[43m    \u001b[49m\u001b[43mreturn_dict\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreturn_dict\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1218\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcache_position\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcache_position\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1219\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1221\u001b[0m hidden_states \u001b[38;5;241m=\u001b[39m outputs[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m   1222\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconfig\u001b[38;5;241m.\u001b[39mpretraining_tp \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/torch/nn/modules/module.py:1532\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1530\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1531\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1532\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/torch/nn/modules/module.py:1541\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1536\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1537\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1538\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1539\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1540\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1541\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1543\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1544\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/transformers/models/llama/modeling_llama.py:1018\u001b[0m, in \u001b[0;36mLlamaModel.forward\u001b[0;34m(self, input_ids, attention_mask, position_ids, past_key_values, inputs_embeds, use_cache, output_attentions, output_hidden_states, return_dict, cache_position)\u001b[0m\n\u001b[1;32m   1007\u001b[0m     layer_outputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_gradient_checkpointing_func(\n\u001b[1;32m   1008\u001b[0m         decoder_layer\u001b[38;5;241m.\u001b[39m\u001b[38;5;21m__call__\u001b[39m,\n\u001b[1;32m   1009\u001b[0m         hidden_states,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1015\u001b[0m         cache_position,\n\u001b[1;32m   1016\u001b[0m     )\n\u001b[1;32m   1017\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1018\u001b[0m     layer_outputs \u001b[38;5;241m=\u001b[39m \u001b[43mdecoder_layer\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1019\u001b[0m \u001b[43m        \u001b[49m\u001b[43mhidden_states\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1020\u001b[0m \u001b[43m        \u001b[49m\u001b[43mattention_mask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcausal_mask\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1021\u001b[0m \u001b[43m        \u001b[49m\u001b[43mposition_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mposition_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1022\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpast_key_value\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpast_key_values\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1023\u001b[0m \u001b[43m        \u001b[49m\u001b[43moutput_attentions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_attentions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1024\u001b[0m \u001b[43m        \u001b[49m\u001b[43muse_cache\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43muse_cache\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1025\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcache_position\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcache_position\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1026\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1028\u001b[0m hidden_states \u001b[38;5;241m=\u001b[39m layer_outputs[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m   1030\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m use_cache:\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/torch/nn/modules/module.py:1532\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1530\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1531\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1532\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/torch/nn/modules/module.py:1541\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1536\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1537\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1538\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1539\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1540\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1541\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1543\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1544\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/transformers/models/llama/modeling_llama.py:741\u001b[0m, in \u001b[0;36mLlamaDecoderLayer.forward\u001b[0;34m(self, hidden_states, attention_mask, position_ids, past_key_value, output_attentions, use_cache, cache_position, **kwargs)\u001b[0m\n\u001b[1;32m    738\u001b[0m hidden_states \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39minput_layernorm(hidden_states)\n\u001b[1;32m    740\u001b[0m \u001b[38;5;66;03m# Self Attention\u001b[39;00m\n\u001b[0;32m--> 741\u001b[0m hidden_states, self_attn_weights, present_key_value \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mself_attn\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    742\u001b[0m \u001b[43m    \u001b[49m\u001b[43mhidden_states\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhidden_states\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    743\u001b[0m \u001b[43m    \u001b[49m\u001b[43mattention_mask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mattention_mask\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    744\u001b[0m \u001b[43m    \u001b[49m\u001b[43mposition_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mposition_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    745\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpast_key_value\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpast_key_value\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    746\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_attentions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_attentions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    747\u001b[0m \u001b[43m    \u001b[49m\u001b[43muse_cache\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43muse_cache\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    748\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcache_position\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcache_position\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    749\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    750\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    751\u001b[0m hidden_states \u001b[38;5;241m=\u001b[39m residual \u001b[38;5;241m+\u001b[39m hidden_states\n\u001b[1;32m    753\u001b[0m \u001b[38;5;66;03m# Fully Connected\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/torch/nn/modules/module.py:1532\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1530\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1531\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1532\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/llm/lib/python3.12/site-packages/torch/nn/modules/module.py:1541\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1536\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1537\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1538\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1539\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1540\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1541\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1543\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1544\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# generate all responses for elife rows and save them in a list\n",
    "# then get rid of all \\n s in the list and save it to elife_llama3_top1_responses.txt\n",
    "elife_responses = []\n",
    "for i in tqdm(range(len(elife_test))):\n",
    "    response = get_laysumm(elife_test, i)\n",
    "    elife_responses.append(response)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open the file in text write mode ('w') instead of binary write mode ('wb')\n",
    "with open('elife_llama3_top1_responses.txt', 'w') as f:\n",
    "    for res in elife_responses:\n",
    "        # Remove all newline characters from the string\n",
    "        res = res.replace('\\n', '')\n",
    "        # Write the modified string to the file, adding a newline character after each string\n",
    "        f.write(res + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 142/142 [20:07<00:00,  8.50s/it]\n"
     ]
    }
   ],
   "source": [
    "plos_responses = []\n",
    "for i in tqdm(range(len(plos_test))):\n",
    "    response = get_laysumm(plos_test, i)\n",
    "    plos_responses.append(response)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open the file in text write mode ('w') instead of binary write mode ('wb')\n",
    "with open('plos_llama3_top1_responses.txt', 'w') as f:\n",
    "    for res in plos_responses:\n",
    "        # Remove all newline characters from the string\n",
    "        res = res.replace('\\n', '')\n",
    "        # Write the modified string to the file, adding a newline character after each string\n",
    "        f.write(res + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
